File: mobile/lib/models/nutrient_deficiency.dart
import 'detection_result.dart';

/// Structured nutrient-deficiency detection response (`NutrientDeficiencyOut`).
///
/// Status values mirror the backend provider layer:
/// - `MODEL_NOT_CONFIGURED` — no model is bundled; never confused with
///   "no deficiency" or a real identification.
class NutrientDeficiencyResult implements DetectionResultModel {
  const NutrientDeficiencyResult({
    required this.success,
    required this.status,
    required this.crop,
    required this.deficiencyName,
    required this.confidence,
    required this.model,
    required this.message,
  });

  static const String statusModelNotConfigured = 'MODEL_NOT_CONFIGURED';

  final bool success;
  final String status;
  final String? crop;
  final String? deficiencyName;
  @override
  final double? confidence;
  final String? model;
  @override
  final String? message;

  @override
  bool get isModelNotConfigured => status == statusModelNotConfigured;

  factory NutrientDeficiencyResult.fromJson(Object? json) {
    final map = json as Map<String, dynamic>;
    return NutrientDeficiencyResult(
      success: map['success'] as bool? ?? true,
      status: map['status'] as String? ?? statusModelNotConfigured,
      crop: map['crop'] as String?,
      deficiencyName: map['deficiency_name'] as String?,
      confidence: (map['confidence'] as num?)?.toDouble(),
      model: map['model'] as String?,
      message: map['message'] as String?,
    );
  }
}

File: config/core/schemas/nutrient_deficiency.py
"""
KisanAI OS
Nutrient Deficiency Detection Schemas
Version: 1.0.0

Future-compatible prediction result. A real model will populate
``deficiency_name``, ``confidence`` and ``model``; until then these
stay ``None`` with ``status = MODEL_NOT_CONFIGURED`` so "no model" is
never confused with "no deficiency" or "deficiency detected". No local
filesystem paths are ever exposed.
"""

from typing import Optional

from pydantic import BaseModel


class NutrientDeficiencyOut(BaseModel):
    """Structured nutrient-deficiency detection response."""

    success: bool = True
    status: str
    crop: Optional[str] = None
    deficiency_name: Optional[str] = None
    confidence: Optional[float] = None
    model: Optional[str] = None
    message: Optional[str] = ""
